x/gamm/pool-models/stableswap/amm_test.go

```diff
@@ -12,6 +13,253 @@ import (
+		uReserve:    osmomath.NewBigDec(200),
+		wSumSquares: osmomath.NewBigDec(20000),
+		yIn:         osmomath.NewBigDec(1),
+		expectPanic: true,
+	},
+	"negative uReserve": {
+		xReserve: osmomath.NewBigDec(100),
+		yReserve: osmomath.NewBigDec(100),
+		// represents a 4-asset pool with 100 in each reserve
+		uReserve:    osmomath.NewBigDec(-200),
+		wSumSquares: osmomath.NewBigDec(20000),
+		yIn:         osmomath.NewBigDec(1),
+		expectPanic: true,
+	},
+	"negative sumSquares": {
+		xReserve: osmomath.NewBigDec(100),
+		yReserve: osmomath.NewBigDec(100),
+		// represents a 4-asset pool with 100 in each reserve
+		uReserve:    osmomath.NewBigDec(200),
+		wSumSquares: osmomath.NewBigDec(-20000),
+		yIn:         osmomath.NewBigDec(1),
+		expectPanic: true,
+	},
+	"negative yIn": {
+		xReserve: osmomath.NewBigDec(100),
+		yReserve: osmomath.NewBigDec(100),
+		// represents a 4-asset pool with 100 in each reserve
+		uReserve:    osmomath.NewBigDec(200),
+		wSumSquares: osmomath.NewBigDec(-20000),
+		yIn:         osmomath.NewBigDec(1),
+		expectPanic: true,
+	},
+	"input greater than pool reserves (even 4-asset pool)": {
+		xReserve:    osmomath.NewBigDec(100),
+		yReserve:    osmomath.NewBigDec(100),
+		uReserve:    osmomath.NewBigDec(200),
+		wSumSquares: osmomath.NewBigDec(20000),
+		yIn:         osmomath.NewBigDec(1000),
+		expectPanic: true,
+	},
+
+	// overflows
+	"xReserve overflows in 4-asset pool": {
+		xReserve: osmomath.NewDecFromBigInt(new(big.Int).Sub(new(big.Int).Exp(big.NewInt(2), big.NewInt(1024), nil), big.NewInt(1))),
+		yReserve: osmomath.NewBigDec(100),
+		// represents a 4-asset pool with 100 in each reserve
+		uReserve:    osmomath.NewBigDec(200),
+		wSumSquares: osmomath.NewBigDec(20000),
+		yIn:         osmomath.NewBigDec(1),
+		expectPanic: true,
+	},
+	"yReserve overflows in 4-asset pool": {
+		xReserve: osmomath.NewBigDec(100),
+		yReserve: osmomath.NewDecFromBigInt(new(big.Int).Sub(new(big.Int).Exp(big.NewInt(2), big.NewInt(1024), nil), big.NewInt(1))),
+		// represents a 4-asset pool with 100 in each reserve
+		uReserve:    osmomath.NewBigDec(200),
+		wSumSquares: osmomath.NewBigDec(20000),
+		yIn:         osmomath.NewBigDec(1),
+		expectPanic: true,
+	},
+	"uReserve overflows in 4-asset pool": {
+		xReserve: osmomath.NewBigDec(100),
+		yReserve: osmomath.NewBigDec(100),
+		// represents a 4-asset pool with 100 in each reserve
+		uReserve:    osmomath.NewDecFromBigInt(new(big.Int).Sub(new(big.Int).Exp(big.NewInt(2), big.NewInt(1024), nil), big.NewInt(1))),
+		wSumSquares: osmomath.NewBigDec(20000),
+		yIn:         osmomath.NewBigDec(1),
+		expectPanic: true,
+	},
+	"wSumSquares overflows in 4-asset pool": {
+		xReserve: osmomath.NewBigDec(100),
+		yReserve: osmomath.NewBigDec(100),
+		// represents a 4-asset pool with 100 in each reserve
+		uReserve:    osmomath.NewBigDec(200),
+		wSumSquares: osmomath.NewDecFromBigInt(new(big.Int).Sub(new(big.Int).Exp(big.NewInt(2), big.NewInt(1024), nil), big.NewInt(1))),
```

Comment:
nit: consider extracting this common "overflow" variable

Reply:
Great point, I also grouped the two asset and multi asset variable declarations together into `var(...)` – looks much cleaner now!